Nodejs/Product/Nodejs/SharedProject/HierarchyIdMap.cs

```diff
@@ -1,25 +1,29 @@
 ﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
+using System.Collections.Concurrent;
 using System.Diagnostics;
 
 namespace Microsoft.VisualStudioTools.Project
 {
     internal sealed class HierarchyIdMap
     {
-        private readonly List<WeakReference<HierarchyNode>> ids = new List<WeakReference<HierarchyNode>>();
-        private readonly Stack<int> freedIds = new Stack<int>();
+        private readonly ConcurrentDictionary<uint, WeakReference<HierarchyNode>> nodes = new ConcurrentDictionary<uint, WeakReference<HierarchyNode>>();
+        private readonly ConcurrentStack<uint> freedIds = new ConcurrentStack<uint>();
 
-        private readonly object theLock = new object();
+        public readonly static HierarchyIdMap Instance = new HierarchyIdMap();
+
+        private HierarchyIdMap() { }
 
         /// <summary>
         /// Must be called from the UI thread
         /// </summary>
         public uint Add(HierarchyNode node)
         {
+            VisualStudio.Shell.ThreadHelper.ThrowIfNotOnUIThread();
+            Debug.Assert(node != null, "The node added here should never be null.");
 #if DEBUG
-            foreach (var reference in this.ids)
+            foreach (var reference in this.nodes.Values)
             {
                 if (reference != null)
```

Comment:
>if (reference != null) [](start = 16, length = 22)

Is this always true?

Reply:
It's kind of weird to do this check by looping, given that they can only be equal if they have the same ID.  If we want to validate the entire structure, it would make more sense to verify that each non-GC'd value has an ID matching its key.

---
In reply to: [182613675](https://github.com/Microsoft/nodejstools/pull/1933#discussion_r182613675) [](ancestors = 182613675)

Reply:
Yes, the weakreference should never be null